src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -399,6 +409,58 @@ bool CanApplyChange(DocumentId documentId)
             }
         }
 
+        private void NotifyRefactorChanges(IEnumerable<DocumentId> changedDocs, CodeAnalysis.Solution oldSolution)
+        {
+            // Without any services to notify there's no need to dig through the documents
+            if (!_refactorNotifyServices.Any())
+            {
+                return;
+            }
+
+            foreach (var changedDocId in changedDocs)
+            {
+                var newDocument = CurrentSolution.GetDocument(changedDocId);
+
+                // Documents without syntax tree won't have the annotations attached
+                if (!newDocument.SupportsSyntaxTree)
+                {
+                    continue;
+                }
+
+                var syntaxRoot = newDocument.GetSyntaxRootSynchronously(CancellationToken.None);
+                var symbolRenameNodes = syntaxRoot.GetAnnotatedNodes(RenameSymbolAnnotation.RenameSymbolKind);
+
+                foreach (var node in symbolRenameNodes)
+                {
+                    foreach (var annotation in node.GetAnnotations(RenameSymbolAnnotation.RenameSymbolKind))
+                    {
+                        var oldDocument = oldSolution.GetDocument(changedDocId);
+
+                        if (!(newDocument.TryGetSemanticModel(out var newSemanticModel) && newSemanticModel is object))
+                        {
+                            continue;
+                        }
+
+                        if (!(oldDocument.TryGetSemanticModel(out var currentSemanticModel) && currentSemanticModel is object))
+                        {
+                            continue;
+                        }
+
+                        var oldSymbol = RenameSymbolAnnotation.ResolveSymbol(annotation, currentSemanticModel.Compilation);
+                        var newSymbol = newSemanticModel.GetSymbolInfo(node).Symbol;
+
+                        foreach (var refactorService in _refactorNotifyServices)
+                        {
+                            if (refactorService.Value.TryOnBeforeGlobalSymbolRenamed(this, changedDocs, oldSymbol, newSymbol.Name, throwOnFailure: false))
+                            {
+                                refactorService.Value.TryOnAfterGlobalSymbolRenamed(this, changedDocs, oldSymbol, newSymbol.Name, throwOnFailure: false);
+                            }
```

Comment:
Do we still need to be raising TryOnAfterGlobalSymbolRenamed _after_ we called base.TryApplyChanges? Like this needs to defer that bit until later?

Reply:
Not sure I follow. This method is called with 

```c#
            if (base.TryApplyChanges(newSolution, progressTracker))
            {
                NotifyRefactorChanges(changedDocs, currentSolution);
                return true;
            }
```

so it should already be done. It's possible we could do the `Before` work ahead, and defer after. I just grouped them all in the same time. We have no resolution if the before fails other than to not call the after. We still apply the change